hilti/toolchain/src/compiler/unit.cc

```diff
@@ -307,6 +307,30 @@ Result<CxxCode> Unit::cxxCode() const {
     return CxxCode{_cxx_unit->moduleID(), cxx};
 }
 
+void Unit::_recursiveDependencies(std::vector<std::weak_ptr<Unit>>* dst, std::vector<const Unit*>* seen) {
+    // This uses two vectors because the weak_ptr are a bit tough to work with,
+    // in particular they can't be compared through std::find().
+
+    for ( const auto& d : _dependencies ) {
+        if ( std::find(seen->begin(), seen->end(), d.lock().get()) != seen->end() )
+            continue;
+
+        dst->push_back(d);
+        seen->push_back(d.lock().get());
+        d.lock()->_recursiveDependencies(dst, seen);
+    }
+}
```

Comment:
Couple of comments:

- I think this would be better as a standalone, hidden function
- a `vector` `seen` seems like an unneeded pessimization for repeated searches, let's use a `unordered_set` instead (which also makes the code for search much shorter).
- let's avoid repeatedly locking the `weak_ptr` as this could become expensive

```suggestion
void _recursiveDependencies(const Unit& unit, std::vector<std::weak_ptr<Unit>>* dst,
                            std::unordered_set<const Unit*>* seen) {
    // This uses two vectors because the weak_ptr are a bit tough to work with,
    // in particular they can't be compared through std::find().

    for ( const auto& d : unit.dependencies(false) ) {
        auto* dep = d.lock().get();
        if ( seen->count(dep) )
            continue;

        dst->push_back(d);
        seen->insert(dep);
        _recursiveDependencies(*dep, dst, seen);
    }
}
```